core/order.go

```diff
@@ -1249,67 +1249,50 @@ func (n *OpenBazaarNode) calculateShippingTotalForListings(contract *pb.Ricardia
 		paymentCurrencyDef, pErr = n.LookupCurrency(paymentCoin)
-		reserveCurrencyDef, rErr = n.LookupCurrency(reserveCurrency)
 	)
 	if oErr != nil {
-		return big.NewInt(0), fmt.Errorf("invalid listing currency code: %s", oErr.Error())
+		if isCryptoListing {
+			// if the currency is unknown assume the default divisibility
+			originCurrencyDef = repo.NewUnknownCryptoDefinition(paymentCoin, repo.DefaultCryptoDivisibility)
+		} else {
+			// Return error if not a valid listing currency code
+			return big.NewInt(0), fmt.Errorf("invalid listing currency code: %s", oErr.Error())
+		}
+
 	}
 	if pErr != nil {
 		return big.NewInt(0), fmt.Errorf("invalid payment currency code: %s", pErr.Error())
 	}
-	if rErr != nil {
-		return big.NewInt(0), fmt.Errorf("invalid reserve currency code: %s", rErr.Error())
-	}
 
+	// No conversion is necessary
 	if originCurrencyDef.Equal(paymentCurrencyDef) {
 		return amount, nil
 	}
 
-	originValue, err := repo.NewCurrencyValue(amount.String(), originCurrencyDef)
-	if err != nil {
-		return big.NewInt(0), fmt.Errorf("parsing amount: %s", err.Error())
-	}
-
-	wal, err := n.Multiwallet.WalletForCurrencyCode(reserveCurrency)
+	reserveWallet, err := n.Multiwallet.WalletForCurrencyCode(reserveCurrency)
 	if err != nil {
 		return big.NewInt(0), fmt.Errorf("%s wallet not found for exchange rates", reserveCurrency)
 	}
 
-	if wal.ExchangeRates() == nil {
-		return big.NewInt(0), ErrPriceCalculationRequiresExchangeRates
-	}
-	reserveIntoOriginRate, err := wal.ExchangeRates().GetExchangeRate(currencyCode)
-	if err != nil {
-		return big.NewInt(0), err
-	}
-	originIntoReserveRate := 1 / reserveIntoOriginRate
-	reserveIntoResultRate, err := wal.ExchangeRates().GetExchangeRate(paymentCoin)
-	if err != nil {
-		// TODO: remove hack once ExchangeRates can be made aware of testnet currencies
-		if strings.HasPrefix(paymentCoin, "T") {
-			reserveIntoResultRate, err = wal.ExchangeRates().GetExchangeRate(strings.TrimPrefix(paymentCoin, "T"))
-			if err != nil {
-				return big.NewInt(0), err
-			}
-		} else {
-			return big.NewInt(0), err
-		}
-	}
+	converter := repo.NewCurrencyConverter(reserveWallet)
 
-	reserveValue, _, err := originValue.ConvertTo(reserveCurrencyDef, originIntoReserveRate)
-	if err != nil {
-		return big.NewInt(0), fmt.Errorf("converting to reserve: %s", err.Error())
-	}
-	resultValue, _, err := reserveValue.ConvertTo(paymentCurrencyDef, reserveIntoResultRate)
+	finalPrice, err := converter.GetFinalPrice(originCurrencyDef, paymentCurrencyDef, amount)
 	if err != nil {
-		return big.NewInt(0), fmt.Errorf("converting from reserve: %s", err.Error())
+		return nil, err
 	}
-	return resultValue.Amount, nil
+
+	// Convert to satoshis
+	divisibility := new(big.Int).SetUint64(uint64(paymentCurrencyDef.Divisibility))
+	divisibilityFactor := new(big.Int).Exp(new(big.Int).SetInt64(10), divisibility, nil)
+	finalPrice.Mul(finalPrice, new(big.Float).SetInt(divisibilityFactor))
+	finalPriceInt, _ := finalPrice.Int(new(big.Int))
```

Comment:
Looks like you do all this as part of `GetFinalPrice`. Let's make sure we don't apply it twice.